src/libstore/gc.cc

```diff
@@ -786,8 +786,18 @@ void LocalStore::collectGarbage(const GCOptions & options, GCResults & results)
         for (auto & i : options.pathsToDelete) {
             assertStorePath(i);
             tryToDelete(state, i);
-            if (state.dead.find(i) == state.dead.end())
-                throw Error(format("cannot delete path '%1%' since it is still alive") % i);
+            if (state.dead.find(i) == state.dead.end()) {
+                /* We cannot delete the current path. It is needed by a
+                 * derivation that cannot be garbage collected. */
+                PathSet incomingLinks;
+                PathSet alreadyVisited;
+                Path gcRootLink;
+                queryReferrers(i, incomingLinks);
+                for (auto & j : incomingLinks)
+                    if(canReachRoot(state, alreadyVisited, i))
+                        gcRootLink = j;
+               throw Error(format("cannot delete path '%1%': it is still needed by '%2%'") % i % gcRootLink);
```

Comment:
Should it be possible for `gcRootLink` to be unassigned here? If yes, we should account for that in the error message. If no, can you add an assert before the `throw`?